pkg/receive/handler.go

```diff
@@ -140,6 +143,9 @@ func (h *Handler) StorageReady() {
 // Hashring sets the hashring for the handler and marks the hashring as ready.
 // If the hashring is nil, then the hashring is marked as not ready.
 func (h *Handler) Hashring(hashring Hashring) {
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+
 	if hashring == nil {
 		atomic.StoreUint32(&h.hashringReady, 0)
```

Comment:
Now that we have introduced a new syncronization mechanism, we can get rid of the hashringReady field, which we were accessing atomically to determine hashring readiness. Instead, `isReady` can do a RLock, e.g.
```go
h.mtx.RLock()
hr := h.Hashring != nil
h.mtx.RUnlock()
return sr > 0 && hr
```